xds/internal/xdsclient/transport/transport.go

```diff
@@ -0,0 +1,564 @@
+	lrsRefCount     int                // Reference count on the load store.
+}
+
+// UpdateHandlerFunc is the implementation at the xDS data model layer, which
+// determines if the configuration received from the management server can be
+// applied locally or not.
+//
+// A nil error is returned from this function when the data model layer believes
+// that the received configuration is good and can be applied locally. This will
+// cause the transport layer to send an ACK to the management server. A non-nil
+// error is returned from this function when the data model layer believes
+// otherwise, and this will cause the transport layer to send a NACK.
+type UpdateHandlerFunc func(update ResourceUpdate) error
+
+// ResourceUpdate is a representation of the configuration update received from
+// the management server. It only contains fields which are useful to the data
+// model layer, and layers above it.
+type ResourceUpdate struct {
+	// Resources is the list of resources received from the management server.
+	Resources []*anypb.Any
+	// URL is the resource type URL for the above resources.
+	URL string
+	// Version is the resource version, for the above resources, as specified by
+	// the management server.
+	Version string
+}
+
+// Options specifies configuration knobs used when creating a new Transport.
+type Options struct {
+	// ServerCfg contains all the configuration required to connect to the xDS
+	// management server.
+	ServerCfg *bootstrap.ServerConfig
+	// UpdateHandler is the component which makes ACK/NACK decisions based on
+	// the received resources.
+	//
+	// Invoked inline and implementations must not block.
+	UpdateHandler UpdateHandlerFunc
+	// StreamErrorHandler provides a way for the transport layer to report
+	// underlying stream errors. These can be bubbled all the way up to the user
+	// of the xdsClient.
+	//
+	// Invoked inline and implementations must not block.
+	StreamErrorHandler func(error)
+	// Backoff controls the amount of time to backoff before recreating failed
+	// ADS streams. If unspecified, a default exponential backoff implementation
+	// is used. For more details, see:
+	// https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md.
+	Backoff func(retries int) time.Duration
+	// Logger does logging with a prefix.
+	Logger *grpclog.PrefixLogger
+}
+
+// For overriding in unit tests.
+var grpcDial = grpc.Dial
+
+// New creates a new Transport.
+func New(opts *Options) (*Transport, error) {
+	switch {
+	case opts == nil:
+		return nil, errors.New("missing options when creating a new transport")
+	case opts.ServerCfg == nil:
+		return nil, errors.New("missing ServerConfig when creating a new transport")
+	case opts.ServerCfg.ServerURI == "":
+		return nil, errors.New("missing server URI when creating a new transport")
+	case opts.ServerCfg.Creds == nil:
+		return nil, errors.New("missing credentials when creating a new transport")
+	case opts.ServerCfg.NodeProto == nil:
```

Comment:
Is a present-but-empty node proto OK?

Also this should be redundant with the following type assertion.

Reply:
Makes sense.

Also: why not make this the proper type in Options since this only accepts one type, though?

Reply:
The node proto is stored in `bootstrap.ServerConfig`. We will change the type in there when getting rid of v2 support.

Reply:
IMO we should push the validation back as close to the source of the data as possible.  I.e. test & convert at the call site instead of making this API accept any type but then validate that it's one specific thing.  WDYT?

Reply:
Oh nevermind, I see what you're saying - this is a nested field.  OK seems fine then.